### PR TITLE
Refactor Application.Run

### PR DESCRIPTION
**Pull request by Gillibald**
## What does the pull request do?
This PR refactors Application in a way that it is closer to WPF's Application. It now has a Startup and Exit event and lets you override OnStartup, OnExit to better customize the lifetime behavior of the application. In addition to that, it is now possible to run an application without showing a window.

## What is the current behavior?
The only way to customize the application startup behavior is to use the AppBuilder and this leads to errors when using controls before the platform is properly initialized.

## What is the updated/expected behavior with this PR?
A customized startup is always running on the UI thread when the platform is already initialized.

## How was the solution implemented (if it's not obvious)?
I have introduced a OnStartup method that can be overridden to customize the startup of an application.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Breaking changes
MainWindow needs to be set explicitly
Application.Exit() -> Application.Shutdown()
Application.ExitMode ->Application.ShutdownMode
Application.OnExiting -> Application.OnExit
Application.OnExit -> Application.Exit
AppBuilderBase<TAppBuilder>.SetExtiMode(ExitMode) -> AppBuilderBase<TAppBuilder>.SetShutdownMode(ShutdownMode)

### CODE REVIEW

**Comment by grokys** [comment]
This looks good in general, and I approve of changing the API to use the same naming as WPF.

I think the one thing to decide (as mentioned in gitter) is how to handle assigning the `MainWindow` when no window is passed to `Run`.

**Comment by grokys** [comment]
Looks good, the only thing that worries me is that we'll be breaking _every single Avalonia application out there_ with the change from `Initialize` -> `OnStartup`. Are we sure that's acceptable? Is there a way we can make applications that don't need any of the functionality in this PR continue to work?

**Comment by kekekeks** [comment]
`OnStartup` and `Initialize` are different. `Initialize` happens *during* initialization process, not everything is ready at that point. `Initialize` is more like auto-generated `InitializeComponent` from WPF.

**Comment by kekekeks** [comment]
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/AppBuilderBase.cs#L308 - there are callbacks that are called after Initialize. If you want `OnStartup`, create a different method.

File: src/Avalonia.Controls/Application.cs [hunk 1]
```diff
@@ -190,18 +194,25 @@ public IResourceDictionary Resources
...
+        ///   <c>true</c> if this instance is shutting down; otherwise, <c>false</c>.
         /// </value>
-        internal bool IsExiting { get; set; }
+        internal bool IsShuttingDown { get; private set; }
 
-        /// <summary>
-        /// Initializes the application by loading XAML etc.
-        /// </summary>
-        public virtual void Initialize()
+        public void Run()
```
Comment (kekekeks) [inline src/Avalonia.Controls/Application.cs]: Right now this method is reentrant. One can stop an app loop and start it again.

I don't see why we are introducing an arbitrary limitation here.

Comment (Gillibald) [reply src/Avalonia.Controls/Application.cs]: When run returns the app closes. How do you start it again then? We could introduce Pause() and Resume(). Maybe I don't understand this entirely.

**Comment by Gillibald** [comment]
Will revert the removal of Application.Initialize

File: src/Avalonia.Controls/Application.cs (hunk 1 above)
Comment (Gillibald) [reply src/Avalonia.Controls/Application.cs]: The limitation should be gone now

**Comment by Gillibald** [comment]
Not sure what is failing here. Maybe some side effects caused by my changes. Trying to reproduce this on my local machine. The weird thing is that platforms fail on different tests.

File: src/Avalonia.Controls/Application.cs [hunk 2]
```diff
@@ -190,108 +194,144 @@ public IResourceDictionary Resources
...
-        internal bool IsExiting { get; set; }
+        internal bool IsShuttingDown { get; private set; }
 
         /// <summary>
         /// Initializes the application by loading XAML etc.
         /// </summary>
-        public virtual void Initialize()
+        public virtual void Initialize() { }
+
+        public int Run()
```
Comment (grokys) [inline src/Avalonia.Controls/Application.cs]: Need an XML doc for this method. Is this correct?

```
/// <summary>
/// Runs the application's main loop until the application is closed by calling <see cref="Shutdown()"/>.
/// </summary>
```

File: src/Avalonia.Controls/IApplicationLifecycle.cs [hunk 3]
```diff
@@ -7,14 +7,25 @@ namespace Avalonia.Controls
...
+        event EventHandler<StartupEventArgs> Startup;
+
         /// <summary>
         /// Sent when the application is exiting.
         /// </summary>
-        event EventHandler OnExit;
+        event EventHandler<ExitEventArgs> Exit;
+
+        /// <summary>
+        /// Shuts down an application that returns the specified exit code to the operating system.
```
Comment (grokys) [inline src/Avalonia.Controls/IApplicationLifecycle.cs]: > that returns the specified exit code to the operating system.

There's no exit code passed to this method, I think this was copy/pasted from the other overload?

File: src/Avalonia.Controls/IApplicationLifecycle.cs [hunk 4]
```diff
@@ -7,14 +7,25 @@ namespace Avalonia.Controls
...
+        /// </summary>
+        void Shutdown();
 
         /// <summary>
-        /// Exits the application.
+        /// Shuts down an application that returns the specified exit code to the operating system.
```
Comment (grokys) [inline src/Avalonia.Controls/IApplicationLifecycle.cs]: I think a better wording might be:

```suggestion
        /// Shuts down the application, returning the specified exit code to the operating system.
```

File: src/Avalonia.Styling/Controls/ExitEventArgs.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) The Avalonia Project. All rights reserved.
+// Licensed under the MIT license. See licence.md file in the project root for full license information.
+
+using System;
+
+namespace Avalonia.Controls
+{
+    public class ExitEventArgs : EventArgs
```
Comment (grokys) [inline src/Avalonia.Styling/Controls/ExitEventArgs.cs]: Could we get some XML docs for this? I'd be happy to provide them if you like.

File: src/Avalonia.Styling/Controls/StartupEventArgs.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) The Avalonia Project. All rights reserved.
+// Licensed under the MIT license. See licence.md file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Avalonia.Controls
+{
+    public class StartupEventArgs : EventArgs
```
Comment (grokys) [inline src/Avalonia.Styling/Controls/StartupEventArgs.cs]: Could we get some XML docs for this? I'd be happy to provide them if you like.

**Review by grokys — CHANGES_REQUESTED** [review]
Getting really close! Just some issues with the XML documentation I think. Let me know if you want me to fix these issues, I'd be happy to do it.

File: src/Avalonia.Controls/IApplicationLifecycle.cs (hunk 3 above)
Comment (Gillibald) [reply src/Avalonia.Controls/IApplicationLifecycle.cs]: I guess this just complicated to get right. The exit code that is set on the event args is set to Environment.ExitCode when this is handled.

**Comment by Gillibald** [comment]
@grokys Would help a lot if you could fix the XML comments. That's something I am struggling a lot.

**Comment by Gillibald** [comment]
@grokys Pls have a look at my recent changes. I have added XML comments. Please let me know if some wording is wrong.

File: src/Avalonia.Controls/Application.cs [hunk 7]
```diff
@@ -190,108 +194,163 @@ public IResourceDictionary Resources
...
 
         /// <summary>
         /// Initializes the application by loading XAML etc.
         /// </summary>
-        public virtual void Initialize()
+        public virtual void Initialize() { }
+
+        /// <summary>
+        /// Runs the application's main loop.
+        /// This will return when the <see cref="Avalonia.ShutdownMode"/> condition is met
```
Comment (grokys) [inline src/Avalonia.Controls/Application.cs]: I think if the summary goes into multiple clauses with descriptions like this, it's probably best to move it to `<remarks>`. So the summary would just be "Run's the application's main loop" and the remarks would explain how the main loop is exited. This makes sense to me because how the application _exits_ isn't really relevant to the summary of the `Run` method.

File: src/Avalonia.Controls/ShutdownMode.cs [hunk 8]
```diff
@@ -6,7 +6,7 @@ namespace Avalonia
     /// <summary>
     /// Enum for ExitMode
```
Comment (grokys) [inline src/Avalonia.Controls/ShutdownMode.cs]:
```suggestion
    /// Describes the possible values for <see cref="Application.ShutdownMode"/>.
```

File: src/Avalonia.Controls/ShutdownMode.cs [hunk 9]
```diff
@@ -6,7 +6,7 @@ namespace Avalonia
...
     /// </summary>
-    public enum ExitMode
+    public enum ShutdownMode
     {
         /// <summary>
         /// Indicates an implicit call to Application.Exit when the last window closes.
```
Comment (grokys) [inline src/Avalonia.Controls/ShutdownMode.cs]:
```suggestion
         /// Indicates an implicit call to Application.Shutdown when the last window closes.
```

File: src/Avalonia.Controls/StartupEventArgs.cs [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+{
+    /// <summary>
+    /// Contains the arguments for the <see cref="IApplicationLifecycle.Startup"/> event.
+    /// </summary>
+    public class StartupEventArgs : EventArgs
+    {
+        private string[] _args;
+
+        /// <summary>
+        /// Gets command line arguments that were passed to the application.
```
Comment (grokys) [inline src/Avalonia.Controls/StartupEventArgs.cs]:
```suggestion
        /// Gets the command line arguments that were passed to the application.
```

**Review by grokys — CHANGES_REQUESTED** [review]
Thanks for that, sorry I didn't have time to do it yet. Just a few tweaks and I think we're good to go.

File: src/Avalonia.Controls/ShutdownMode.cs (hunk 8 above)
Comment (Gillibald) [reply src/Avalonia.Controls/ShutdownMode.cs]: ExitMode - > ShutdownMode

File: src/Avalonia.Controls/Application.cs [hunk 11]
```diff
@@ -190,108 +194,163 @@ public IResourceDictionary Resources
...
+            IsShuttingDown = true;         
 
             Windows.Clear();
 
-            OnExit?.Invoke(this, EventArgs.Empty);
+            try
+            {
+                var e = new ExitEventArgs { ApplicationExitCode = _exitCode };
+
+                OnExit(e);
```
Comment (kekekeks) [inline src/Avalonia.Controls/Application.cs]: `OnExit` naming is inconsistent with `Shutdown`

File: src/Avalonia.Controls/Application.cs [hunk 12]
```diff
@@ -190,108 +194,163 @@ public IResourceDictionary Resources
...
+
+                _exitCode = e.ApplicationExitCode;
+
+                Environment.ExitCode = _exitCode;
```
Comment (kekekeks) [inline src/Avalonia.Controls/Application.cs]: Also `_exitCode`, but `IsShuttingDown`

**Comment by kekekeks** [comment]
Also, just a note: all of that shutdown stuff will only work on several platforms (desktop ones). On most of user devices there is no event loop we can control.

File: src/Avalonia.Controls/Application.cs (hunk 11 above)
Comment (Gillibald) [reply src/Avalonia.Controls/Application.cs]: OnExit raises an event and shutdown starts the process of shutting down. I don't get your point. Pls explain this further.

File: src/Avalonia.Controls/Application.cs (hunk 12 above)
Comment (Gillibald) [reply src/Avalonia.Controls/Application.cs]: Exit is the point in time when everything stops. Shutting down is a range in time. They don't need to be named the same because they are not the same.

Comment (grokys) [reply src/Avalonia.Controls/Application.cs]: This is consistent with WPF:

https://docs.microsoft.com/en-us/dotnet/api/system.windows.application.shutdown?view=netframework-4.8
https://docs.microsoft.com/en-us/dotnet/api/system.windows.application.onexit?view=netframework-4.8
https://docs.microsoft.com/en-us/dotnet/api/system.windows.application.exit?view=netframework-4.8

**Comment by Gillibald** [comment]
Is there a reason we don't spawn our own ui thread on mobile? Avalonia's ui is independent of the platform's ui. I don't like how mobile and desktop diverge. Application should be the main entry point and be responsible for the lifetime.

**Comment by kekekeks** [comment]
>Is there a reason we don't spawn our own ui thread on mobile? 

TL;DR: You can't

**Review by grokys — APPROVED** [review]

**Comment by danwalmsley** [comment]
This has somehow broken OSX. We need to a some kind of unit test that checks if control catalog runs on each platform.

**Comment by Gillibald** [comment]
What is exactly failing?
